docs/design/proposals/cephfs-fscrypt.md

```diff
@@ -0,0 +1,281 @@
+# Ceph Filesystem fscrypt Support
+
+## Problem Description
+
+As a Ceph Container Storage Interface (Ceph CSI) user, I want a cloud native way
+to manage keys and enable encryption on Ceph Filesystem (CephFS) volumes.
+
+In order to access encrypted volumes without Ceph CSI, this can be done by unlocking
+volumes with user space tools.
+
+## Background
+
+*fscrypt* or *FSCrypt* is a Linux Kernel feature that allows the filesystem
+to support the transparent encryption of files and directories. Local
+filesystems like ext4 and F2FS (Flash-Friendly File System) support this feature
+already.
+
+Work is in progress to add fscrypt support to CephFS for filesystem-level encryption.
+
+ - [FSCrypt Kernel Documentation](https://www.kernel.org/doc/html/latest/filesystems/fscrypt.html)
+ - Management Tools
+   - [`fscrypt`](https://github.com/google/fscrypt)
+   - [`fscryptctl`](https://github.com/google/fscryptctl)
+ - [Ceph Feature Tracker: "Add fscrypt support to the kernel CephFS client"](https://tracker.ceph.com/issues/46690)
+ - [`fscrypt` design document](https://goo.gl/55cCrI)
+
+**NOTE:** In this document, fscrypt refers to the filesystem-level encryption
+feature, while `fscrypt` specifically refers to the user space tool.
+
+## Terminology
+ - *FSCrypt*, *fscrypt* - Linux Kernel filesystem-level encryption feature
+ - `fscrypt` (code formatted) - User space tool manage keys and encryption policies
+ - `fscryptctl` (code formatted) - Low-level user space tool manage keys and encryption policies
+ - *subvolume* - CephFS subvolume
+ - *unlocking* - Using a key to make an encrypted filesystem
+   accessible in plain text
+ - *protector* (fscrypt) - A single method or secret plus data used to
+   derive a *protector key*. Example: user login passphrase
+ - *protector key* (fscrypt) - A symmetric key derived from an external
+   source. Used by a *policy* to unwrap a *policy key*
+ - *policy key* (fscrypt) - An encryption key passed to the kernel to
+   unlock a directory
+ - *policy* (fscrypt) - A collection of directories protected and unlocked as a unit
+ - *KMS* - Key management system
+
+## User Visible Change
+
+Similar to the existing RADOS Block Device (RBD) encryption support,
+we propose adding encryption support in the configuration and Key Management Service (KMS)
+integration.
+
+In this example, a user may enable encryption using storage class keys similar
+to RBD. Ceph CSI then configures and unlocks the persistent volumes and CephFS
+subvolumes.
+
+Due to the way `fscrypt` stores metadata, subvolumes have a regular root
+directory containing  a `/.fscrypt` directory and a
```

Comment:
FWIW, the `.fscrypt` directory is one of the main reasons I wanted the CSI team to think about this. This directory is not generally an issue with local filesystems like ext4, as admins almost always mount the root of the fileystem, and you can generally ensure that this directory is available to userland applications.

Contrast that with something like NFS or Ceph, where mounting a subdirectory of an export is very common. You could end up in a situation where you've run `fscrypt setup` on an upper-level directory but then someone mounts a subdirectory of it. The `.fscrypt` directory won't be available on the client at that point.

The placement of the .fscrypt directory is crucial if you intend to use the fscrypt binary under the hood. I wonder if it may even be better to do something like store the info that's currently in `.fscrypt` directly in a RADOS objects instead, which would sort of sidestep the whole issue. That would mean a major overhaul for fscrypt, however (or a rewrite).

Reply:
> The problem would be access control in the spirit of fscrypt. /.fscrypt is basically set up like /tmp (sticky). Created by root, but non-root users are allowed to add/remove their own protectors.

Good point. That would be tricky to replicate in bare RADOS with cephx creds. You'd probably have to layer some enforcement on top. Blech.

That said...one thing that the local fscrypt folks don't need to worry about is idmapping. I imagine consistent uid/gid mapping is a must for k8s nodes so I'll assume that's not a problem here.

> We could move the .fscrypt directory as high as the filesystem or the csi subvolume group. Basically having it next to the "meta" files.

The key point here is that the `.fscrypt` dir has to be consistently reachable, so you need to consistently mount the same cephfs directory. If you're dealing with a pretty flat hierarchy where every tenant is mounting his own subvolume, you should be ok. If, on the other hand, you have a situation where some clients mount at a higher point in the directory tree then things get more iffy.

I imagine Ceph CSI keeps things fairly flat though, so you should be OK.

Having multiple fscrypt directories (one for each subvolume) seems like the best approach, IMO. Allowing access to the keys in there should be safe (since they are just part of the overall KDF), but it'd be better to deny access to it when tenants don't need it.